### PR TITLE
Generic prefab loading from prefab folder

### PR DESCRIPTION
**Pull request by sunreef**
Added generic prefab loading from prefab folder. Prefabs are now stored with the name of the entity in a `HashMap`.

Spawn events have been changed to use a `String` instead of an enum to determine which creature to spawn.

### CODE REVIEW

File: resources/prefabs/bee.ron [hunk 1]
```diff
@@ -0,0 +1,35 @@
+#![enable(implicit_some)]
```
Comment (marotili) [inline resources/prefabs/bee.ron]: I think it would be great if you can split the bee into a separate pull request so that we can refer people to it if they ask how to add creatures.

File: src/resources/prefabs.rs [hunk 2]
```diff
@@ -1,48 +1,96 @@
...
+    pub fn get_prefabs(&self) -> &HashMap<String, Handle<Prefab<CreaturePrefabData>>> {
+        &self.prefabs
+    }
+
+    pub fn set_prefabs(&mut self, prefabs: HashMap<String, Handle<Prefab<CreaturePrefabData>>>) {
+        self.prefabs = prefabs;
+    }
 }
 
 // Here we load all prefabs for the different creatures in the game.
```
Comment (marotili) [inline src/resources/prefabs.rs:39]: Can you update the documentation for `initialize_prefabs` and `update_prefabs`, since it is using a workaround?

File: src/resources/prefabs.rs [hunk 3]
```diff
@@ -1,48 +1,96 @@
...
-                loader.load("prefabs/herbivore.ron", RonFormat, (), ()),
-                loader.load("prefabs/plant.ron", RonFormat, (), ()),
-            )
-        });
-    creature_prefabs.insert("Carnivore".to_string(), carnivore_prefab);
-    creature_prefabs.insert("Herbivore".to_string(), herbivore_prefab);
-    creature_prefabs.insert("Plant".to_string(), plant_prefab);
+    let mut progress_counter = ProgressCounter::new();
+    let prefab_iter = {
+        let prefab_dir_path = application_root_dir() + "/resources/prefabs";
```
Comment (marotili) [inline src/resources/prefabs.rs]: Please move the creatures into a separate subfolder, in case we add other prefabs

File: src/resources/prefabs.rs [hunk 4]
```diff
@@ -1,48 +1,96 @@
...
+                    RonFormat,
+                    (),
+                    &mut progress_counter,
+                )
+            })
+        })
+    };
+
+    let mut count = 0;
+    for prefab in prefab_iter {
```
Comment (marotili) [inline src/resources/prefabs.rs]: you could use `for (index, prefab) in prefab_iter.enumerate()` here I think, see https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.enumerate

File: src/resources/prefabs.rs [hunk 5]
```diff
@@ -1,48 +1,96 @@
...
+    }
     world.add_resource(creature_prefabs);
+    progress_counter
+}
+
+pub fn update_prefabs(world: &mut World) {
+    let updated_prefabs = {
+        let creature_prefabs = world.read_resource::<CreaturePrefabs>();
+
+        let prefabs = creature_prefabs.get_prefabs().clone();
```
Comment (marotili) [inline src/resources/prefabs.rs]: Is this clone necessary?

File: src/resources/prefabs.rs [hunk 6]
```diff
@@ -1,48 +1,96 @@
...
+        let mut prefab_resource =
+            world.write_resource::<AssetStorage<Prefab<CreaturePrefabData>>>();
+        let mut new_prefabs = HashMap::new();
+        for (_key, handle) in prefabs.iter() {
+            let prefab = prefab_resource.get_mut(handle).unwrap();
```
Comment (marotili) [inline src/resources/prefabs.rs]: Are those unwraps safe?

File: src/states/main_game.rs [hunk 7]
```diff
@@ -204,10 +185,47 @@ impl<'a> State<GameData<'a, 'a>, CustomStateEvent> for MainGameState {
 
     fn update(
         &mut self,
-        data: StateData<'_, GameData<'a, 'a>>,
+        mut data: StateData<'_, GameData<'a, 'a>>,
     ) -> Trans<GameData<'a, 'a>, CustomStateEvent> {
         self.dispatcher.dispatch(&mut data.world.res);
         data.data.update(&data.world);
+
+        match self.prefab_loading_progress.as_ref() {
```
Comment (marotili) [inline src/states/main_game.rs]: Replace by `if let Some(counter) = ...`

File: src/systems/spawner.rs [hunk 8]
```diff
@@ -77,18 +77,26 @@ impl<'s> System<'s> for CreatureSpawnerSystem {
...
+        Write<'s, EventChannel<CreatureSpawnEvent>>,
+        Read<'s, Time>,
+        Read<'s, CreaturePrefabs>,
+    );
 
-    fn run(&mut self, (mut spawn_events, time): Self::SystemData) {
+    fn run(&mut self, (mut spawn_events, time, _creature_prefabs): Self::SystemData) {
         let delta_seconds = time.delta_seconds();
         self.timer_to_next_spawn -= delta_seconds;
+        self.bee_timer -= delta_seconds;
```
Comment (marotili) [inline src/systems/spawner.rs]: As mentioned above, it would be great if you could move this into a separate PR. In addition, you could write a separate system so that we can enable/disable bee spawning by registering/unregistering systems.

**Review by marotili — APPROVED** [review]
Very nice 👌 I have a couple of comments but LGTM in general :)

File: src/systems/spawner.rs (hunk 8 above)
Comment (sunreef) [reply src/systems/spawner.rs]: I removed the bee from this PR.

**Review by marotili — APPROVED** [review]
Please also remove `tiny-bee.png`.
Can you resolve all comments that you addressed, or respond to them if you decided not to apply the suggested changes?

File: src/resources/prefabs.rs (hunk 2 above)
Comment (sunreef) [reply src/resources/prefabs.rs:39]: Which documentation do you mean ? The dev log ? Or just comments in the code ?

Comment (marotili) [reply src/resources/prefabs.rs:39]: Just comments in the code, so that others browsing the code will know about the relationship between initialize and update
